qa/suites/upgrade/octopus-x/stress-split/2-partial-upgrade/firsthalf.yaml

```diff
@@ -3,12 +3,11 @@ meta:
    install upgrade ceph/-x on 2/3 of cluster
    restart : mons, osd.0-7
 tasks:
-- install.upgrade:
-    mon.a:
-    mon.b:
-    mon.c:
+- cephadm.shell:
+    env: [sha1]
+    #we need to upgrade only nodes a b c - how? syntax or else?
```

Comment:
the upgrade has a specific order:

https://github.com/ceph/ceph/blob/c3af7f4fb7357be51d61cb154046587e550b8b29/src/pybind/mgr/cephadm/upgrade.py#L17

we first upgrade all MGRs, then we upgrade all MONs, then we upgrade all OSDs. etc. this is not by host, but by service type. What do you want? do you want to have some OSDs upgraded? that should work by doing something like:

```python
def is_split_good() -> bool:
    ps = check_output("ceph orch ps --format json")
    if ... 

while not is_split_good():
    time.sleep(5)

Reply:
@sebastian-philipp In the existing stress-split upgrade suites like https://github.com/ceph/ceph/tree/master/qa/suites/upgrade/nautilus-x/stress-split, the split is created by upgrading the mgr, few mons and osds first, running some workloads and finishing the upgrade with mon, osds and clients. Since cephadm defines `CEPH_UPGRADE_ORDER`, which I believe is not being overriden anywhere, does the existing stress-split structure still make sense? or should we create a split by abiding by the `CEPH_UPGRADE_ORDER` defined by cephadm.

Reply:
maybe it would make more sense to manually update a few daemons. Like providing a mon command

```
ceph orch redeploy mgr.x --image quay.ceph.io/ceph-ci/ceph:$sha1
```

and then do something like

```yaml
tasks:
- cephadm.shell:
  host.a:
  - ceph orch redeploy mon.a --image quay.ceph.io/ceph-ci/ceph:$sha1
  - ceph orch redeploy mgr.x --image quay.ceph.io/ceph-ci/ceph:$sha1
  - ceph orch redeploy osd.0 --image quay.ceph.io/ceph-ci/ceph:$sha1
  - ceph orch redeploy ceph.rgw.realm.zone.a --image quay.ceph.io/ceph-ci/ceph:$sha1

# do some tests ...

# and then finally finish the upgrade:
  - ceph orch upgrade start --image quay.ceph.io/ceph-ci/ceph:$sha1
```

and then completely ignore the upgrade order for this test.

Reply:
@sebastian-philipp I think it makes sense to follow the upgrade order in these tests since that's what our users will be using. We just need the ability to reliably stop the upgrade at any point while running workloads in parallel. Ideally, we'll want to cover test cases where we have mixed versions of mgrs, mons and osds. I presume this will need some changes to the upgrade module. 
@jdurgin @yuriw makes sense?

Reply:
We have conflicting goals here:

* cephadm tries to provide an upgrade that is as safe as possible by sticking to a pre-defined upgrade order leveraging the benefits of a container based deployment.
* The upgrade test tries to reveal hidden bugs when running ceph in awkward upgrade scenarios. Which is exactly what cepham tries to avoid.

I mean, we not only need to modify the upgrade order, but also transpose it. (e.g. instead of 

* `mgr.x, mgr.y, mgr.z, mon.a, mon.b, mon.c, osd.0, osd.1, osd.2`

change it to something similar to 

* `mgr.x, mon.a, osd.0, mgr.y, mon.b, osd.1, mgr.z, mon.c, osd.2`

)

And this non-advisable upgrade order should be implemented in qa/ and not in src/.

Reply:
Let me explain again.

> We have conflicting goals here:
> 
> * cephadm tries to provide an upgrade that is as safe as possible by sticking to a pre-defined upgrade order leveraging the benefits of a container based deployment.

I think we should stick to the pre-defined order in our tests too. So the stress-split changes in this PR are not relevant.

> * The upgrade test tries to reveal hidden bugs when running ceph in awkward upgrade scenarios. Which is exactly what cepham tries to avoid.


What I am suggesting is to redesign our test suites such that they create awkward upgrade scenarios but do not defy the upgrade order. The first half of the split could look like

* `mgr.x, mgr.y`

or

* `mgr.x, mgr.y, mgr.z, mon.a, mon.b`

or 

* `mgr.x, mgr.y, mgr.z, mon.a, mon.b, mon.c, osd.0, osd.1, osd.2`


> 
> I mean, we not only need to modify the upgrade order, but also transpose it. (e.g. instead of

No, we should not transpose anything.

> 
> * `mgr.x, mgr.y, mgr.z, mon.a, mon.b, mon.c, osd.0, osd.1, osd.2`
> 
> change it to something similar to
> 
> * `mgr.x, mon.a, osd.0, mgr.y, mon.b, osd.1, mgr.z, mon.c, osd.2`
> 
> )
> 
> And this non-advisable upgrade order should be implemented in qa/ and not in src/.

Reply:
For now, I can offer you this with #36426 :

```
echo mgr.x mgr.y mgr.z mon.a mon.b mon.c osd.0 | xargs -n 1 ceph orch daemon redeploy --image=quay.cepi.io/ceph-ci/ceph:$sha --name 
```

Having something more sophisticated in mgr/cephadm will take some time.